src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -153,24 +167,126 @@ void GcsPlacementGroupManager::HandleCreatePlacementGroup(
+    if (status.ok()) {
+      RAY_LOG(INFO) << "Placement group of an id, " << placement_group_id
+                    << " is removed successfully.";
+    } else {
+      RAY_LOG(INFO) << "Removing a placement group of an id, " << placement_group_id
+                    << " failed.";
+    }
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  });
+}
+
+void GcsPlacementGroupManager::RemovePlacementGroup(
+    const PlacementGroupID &placement_group_id,
+    StatusCallback on_placement_group_removed) {
+  RAY_CHECK(on_placement_group_removed);
+  // If the placement group is not registered, it means it is already removed or wasn't
+  // created before. Just fail the request in this case.
+  if (registered_placement_groups_.find(placement_group_id) ==
+      registered_placement_groups_.end()) {
+    RAY_CHECK(registered_placement_groups_.find(placement_group_id) ==
+              registered_placement_groups_.end());
+    std::stringstream stream;
+    stream << "Placement group of id " << placement_group_id
+           << " is already removed or doesn't exist";
+    on_placement_group_removed(Status::NotFound(stream.str()));
+    return;
+  }
+
+  auto created_placement_group_it = created_placement_group_.find(placement_group_id);
+  if (created_placement_group_it != created_placement_group_.end()) {
+    // If the placement group is already created.
+    gcs_placement_group_scheduler_->DestroyPlacementGroupResources(placement_group_id);
+    created_placement_group_.erase(created_placement_group_it);
+  } else if (IsSchedulingInProgress(placement_group_id)) {
+    // If the placement group is scheduling.
+    gcs_placement_group_scheduler_->CancelScheduling(placement_group_id);
+  } else {
+    // If placement group is pending
+    auto pending_it = std::find_if(
+        pending_placement_groups_.begin(), pending_placement_groups_.end(),
+        [placement_group_id](const std::shared_ptr<GcsPlacementGroup> &placement_group) {
+          return placement_group->GetPlacementGroupID() == placement_group_id;
         });
+
+    if (pending_it != pending_placement_groups_.end()) {
+      // The placement group was pending scheduling, remove it from the queue.
+      pending_placement_groups_.erase(pending_it);
+    } else {
+      // This can happen only when placement group is removed as soon as it is created.
+      // TODO(sang): Make the initial registration synchronous, so that we don't need to
+      // care about this edge case.
+      RAY_CHECK(registered_placement_groups_.find(placement_group_id) !=
+                registered_placement_groups_.end());
+    }
+  }
+
+  auto placement_grouop_it = registered_placement_groups_.find(placement_group_id);
+  auto placement_group = placement_grouop_it->second;
+  registered_placement_groups_.erase(placement_grouop_it);
+  // If placement group hasn't been created yet, send a response to a core worker that
```

Comment:
`OnPlacementGroupResourceReserveDone` will call the callback if we remove placement group before it is created?

Reply:
`placement_group_to_register_callback_` is not called because that will always call the failure callback. But, I think I should add logic inside `OnPlacementGroupCreationFailed`, otherwise it will add the removed placement group to the pending list again.